src/controllers/csv-processor.ts

```diff
@@ -18,6 +21,21 @@ export const MAX_PAGE_SIZE = 500;
 export const MIN_PAGE_SIZE = 5;
 export const DEFAULT_PAGE_SIZE = 100;
 
+function hashReadableStream(stream: Readable, algorithm: string = 'sha256'): Promise<string> {
```

Comment:
At some stage we have to read the stream if you want hashes of files... In terms of order... I've left hashing to after we've sent the file up to Azure Blob Storage for this reason.  The alternative is we abandon the concept has hashing and storing hashes and ensuring integrality.  Which would make things simpler but equally we can't do you other request which is to prevent storage of duplicate files within the datalake.  There are trade offs.

Reply:
See how this is done in spider.py / ugc.py.
We never force materialisation of the stream. We batch it into the stream and peek at it along the way, maintaining small, fixed sized state in the digest object.
So yes, we have to read it, but there's no need to ever materialise the whole stream.